rfd/0163-vnet.md

```diff
@@ -0,0 +1,796 @@
+another VPN client is being used that happens to use the same range.
+For example, tailscale also uses this range.
+This range can be overriden with a cluster-wide configuration.
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  cidr_range: 100.64.0.0/10
+```
+
+#### IPv6
+
+IPv6 addresses will be assigned according to RFC 4193 "Unique Local IPv6 Unicast
+Addresses".
+In accordance with the RFC, VNet will randomly generate a 40-bit global ID on
+startup which will be used as a prefix for all IPs assigned to apps and to
+assign an IP route to the TUN interface.
+The CIDR range will not be configurable for IPv6.
+
+#### IP access without DNS
+
+If a client wishes to access a Teleport app over VNet but their client does not
+accept a domain name and only an IP, there are two ways this can be handled.
+
+The first option is the user can simply ping to domain name to cause VNet to
+assign an IP to the app, and then they may use that IP.
+
+```shell
+$ ping api.teleport.example.com
+PING api.teleport.example.com (100.64.0.4): 56 data bytes
+$ curl 100.64.0.4
+```
+
+Another option would be to support static configuration of an IP that will
+always be assigned to an app.
+This may be desirable by some users used to accessing an app at a static IP, but
+the implementation would have many corner cases and could be fragile so we would
+prefer not to support this in the initial implementation.
+
+### TCP ports
+
+Local app proxies today each listen on a unique TCP port on `127.0.0.1`.
+With VNet we will assign a unique IP address to each app with the full TCP port
+range available.
+
+The reason we need to assign an IP per app, instead of having many apps on the
+same IP but different ports, is that DNS only assigns an IP to each domain name
+with no port, so there would be no good way for the client to discover the port
+we assigned to the app.
+
+To keep configuration as simple as possible, and because we can, the default
+behavior of VNet will be to proxy TCP connections from ANY port to the app
+assigned to the IP.
+This way clients can use the port they expect for the app protocol, or any port
+they like, with no extra configuration or need to figure out exactly which port
+to use.
+
+The exception to this is if the configured `vnet_addr` for the app explicitly
+includes a port then VNet will only handle connections on that port.
+This will allow multiple Teleport apps to be configured on different ports at
+the same address.
+For example, the following configuration would make both ports 80 and 443
+available for `app.example.com`.
+VNet would assign the same IP address to both of these apps because they have
+the same `vnet_addr` only differing by port.
+
+```yaml
+kind: app
+metadata:
+  name: https-app
+spec:
+  uri: tcp://127.0.0.1:443
+  vnet_addr: app.example.com:443
+---
+kind: app
+metadata:
+  name: http-app
+spec:
+  uri: tcp://127.0.0.1:80
+  vnet_addr: app.example.com:80
+```
```

Comment:
I suppose this is no longer relevant now that `spec.vnet_addr` is gone.

Reply:
Yeah forgot to remove this. It might have been nice to have two apps on different ports of the same address but I think it's better to simplify the dns names and configuration. If we really want this feature later we can always add it